### PR TITLE
Lab 3

### LINKED ISSUE #295: Lab 3
- Complete Lab 3
 - letters function
 - sortLetters function
- Add Unit Tests for 100% test coverage

### PR DESCRIPTION
**Pull request by joshcarp**
Fixes #295 

Review of colleague's PR #

#### Changes proposed in this PR:

- Complete Lab 3
 -  Add letters function
 -  Add sortLetters function
- Add Unit Tests for 100% test coverage

### CODE REVIEW

File: 03_letters/joshcarp/letters.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+func letters(s string) map[rune]int {
+	freq := make(map[rune]int)
+	for _, i := range s {
+		freq[i]++
+	}
+	return freq
+}
+
+func sortLetters(m map[rune]int) []string {
+	var ans StringSlice
```
Comment (golangcibot) [inline 03_letters/joshcarp/letters.go]: Consider preallocating `ans` (from `prealloc`)

File: 03_letters/joshcarp/letters_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+	return r, w
+}
+
+// captureStop copies a file buffer and returns a string of the file
+func captureStop(r *os.File, w *os.File) string {
+	var buf bytes.Buffer
+	w.Close()
+	io.Copy(&buf, r)
```
Comment (golangcibot) [inline 03_letters/joshcarp/letters_test.go]: Error return value of `io.Copy` is not checked (from `errcheck`)

File: 03_letters/joshcarp/letters_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+// captureStart diverts stdio to another file object
+func captureStart() (*os.File, *os.File) {
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+	return r, w
+}
+
+// captureStop copies a file buffer and returns a string of the file
+func captureStop(r *os.File, w *os.File) string {
```
Comment (golangcibot) [inline 03_letters/joshcarp/letters_test.go]: `r` can be `io.ReadCloser` (from `interfacer`)

**Review by jonathanheaden — APPROVED** [review]
Looks good.
One observation that was made to me on a prior submission is that you might get better use of  `assert`  rather than `require` in the testing package. `Assert` will not terminate failure the way `require` does

File: 03_letters/joshcarp/letters.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type StringSlice []string
```
Comment (rokane) [inline 03_letters/joshcarp/letters.go:9]: I think you might be doing this so that you can use the `sort.Sort` function, but I don't think creating a custom type which is just a wrapper around a normal type simplifies the solution. 

You should be able to use a normal `[]string` and use the `sort.Slice` function to achieve the same outcome. This will simplify your solution and you won't need this custom wrapper code.

File: 03_letters/joshcarp/letters_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+}{
+	{"aba", map[rune]int{'a': 2, 'b': 1}, []string{"a:2", "b:1"}},
+	{"ywbgjzhoiumsnqdexprlctvfka", map[rune]int{
+		'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1, 'g': 1, 'h': 1, 'i': 1,
+		'j': 1, 'k': 1, 'l': 1, 'm': 1, 'n': 1, 'o': 1, 'p': 1, 'q': 1, 'r': 1,
+		's': 1, 't': 1, 'u': 1, 'v': 1, 'w': 1, 'x': 1, 'y': 1, 'z': 1},
+		[]string{"a:1", "b:1", "c:1", "d:1", "e:1", "f:1", "g:1", "h:1", "i:1",
+			"j:1", "k:1", "l:1", "m:1", "n:1", "o:1", "p:1", "q:1", "r:1",
+			"s:1", "t:1", "u:1", "v:1", "w:1", "x:1", "y:1", "z:1"}},
+	{`.#jV#Rz;ZO`, map[rune]int{
```
Comment (rokane) [inline 03_letters/joshcarp/letters_test.go:28]: What about a couple more edge cases?

Empty input?
Non ASCII characters (😀😃) to test that a rune map is being used as the underlying datastore?

File: 03_letters/joshcarp/letters_test.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		'#': 2, '.': 1, ';': 1, 'O': 1, 'R': 1, 'V': 1, 'Z': 1, 'j': 1, 'z': 1},
+		[]string{"#:2", ".:1", ";:1", "O:1", "R:1", "V:1", "Z:1", "j:1", "z:1"}},
+}
+
+func TestMain(t *testing.T) {
```
Comment (rokane) [inline 03_letters/joshcarp/letters_test.go:33]: Nice work 👍

**Review by rokane — CHANGES_REQUESTED** [review]
Really nice work. I like the way you structured your tests and used the one `lettersTests` struct to hold information for both parts of the solution. I think you can simplify the solution a little more by removing the custom `StringSlice` data type, and possibly add a few more extra test cases. Other than that, everything looks super clean and really went done :)

**Comment by juliaogris** [comment]
The go-course is now closed. Thank you very much for participating.
